Feature/admin paginated users

EmCousin opened the pull request:
# Optimize Admin Refund Queue with Pagination

- [x] Merge https://github.com/antiwork/gumroad/pull/1804 first

## AI Disclosure

- Pagination feature was coded manually in #1586, refactored manually and reviewed with AI 

## Problem

The admin refund queue page was experiencing severe performance issues when loading users:
- **Loading time**: ~3 seconds to load just 15 users (up to 5 seconds before we migrated the refund queue to Inertia)
- **User experience**: Long wait times before any content appeared
- **Scalability**: Would become unusable with larger datasets

The root cause was loading all users at once with their associated data (avatars, memberships, links, purchases), which created expensive database queries and large payload transfers.

## Solution

Implemented aggressive pagination with infinite scroll to dramatically improve initial page load time and user experience. Since each user card takes substantial vertical space on the page, users naturally take time scrolling through results, making infinite scroll a suitable UX pattern for this use case.

### Key Changes

#### 1. Backend Pagination with Pagy
**Commit:** `5c976b9e`

- Created reusable `Admin::ListPaginatedUsers` concern
- Integrated `pagy_countless` for efficient "load more" pagination
  - Uses `countless_minimal: true` to avoid expensive COUNT queries
  - Configurable page size (default: 25 records per page)
- Leveraged `InertiaRails.merge` to accumulate users across page loads
- Returns both `users` array and `pagination` object to frontend

**Why:** `pagy_countless` is optimized for infinite scroll scenarios - it doesn't run expensive `COUNT` queries and only checks if there's a next page by fetching N+1 records.

#### 2. Frontend Infinite Scroll with Inertia's `WhenVisible`component
**Commit:** `198a3d38`

- Created new `PaginatedLoader` component using Inertia's `<WhenVisible>`
- Automatically triggers next page load when component becomes visible in viewport
- Displays loading indicator while fetching
- Stops loading when last page returns fewer items than limit
- Added unique key `${pagination.page}-${pagination.limit}` to force remounting on page changes

**Key features:**
```tsx
// Determines if we should load more by checking if current page was full
const expectedItemsUpToCurrentPage = pagination.page * pagination.limit;
const hasFullPage = itemsLength >= expectedItemsUpToCurrentPage;
```

**Why:** Intersection Observer-based infinite scroll provides the best UX for this use case - users see content immediately and additional pages load seamlessly as they scroll.

#### 4. Fix Inertia Route Matching with Query Params
**File:** `app/javascript/packs/admin_inertia.ts`

- Updated route matching to strip query parameters before checking if URL has migrated to Inertia
- Changed from `url` to `url.split("?")[0]` for comparison

**Why:** Pagination adds `?page=2` query params, which broke the Inertia migration detection and caused full page reloads.

#### 5. Deterministic Ordering for Pagination
**Commit:** `d7e6edb0`

- Added `id: :desc` as secondary sort to `User.refund_queue` ordering
- Changed from `.order(updated_at: :desc)` to `.order(updated_at: :desc, id: :desc)`

**Why:** Without a deterministic sort (secondary key), pagination can return duplicate or missing records when multiple users have the same `updated_at` timestamp. Adding `id` as a tiebreaker ensures consistent, predictable pagination. Unlikely to happen in production, it was systematically the case in development.

## Performance Impact

### Before
- Initial load: ~2,7 seconds for 13 users
- All users loaded at once
- Long blank page while waiting

<img width="1316" height="883" alt="Screenshot 2025-10-23 at 15 01 04" src="https://github.com/user-attachments/assets/8798b0f3-5b0c-40d8-be42-4fd76121138c" />

### After
- Initial load: ~300-400ms for first 5 users
- Subsequent pages: Load seamlessly as user scrolls
- Content visible immediately
- Scales efficiently to any number of users in queue

<img width="1313" height="883" alt="Screenshot 2025-10-23 at 14 57 49" src="https://github.com/user-attachments/assets/06ec0af6-2153-457f-8857-dca40973bf91" />

https://github.com/user-attachments/assets/f49b0b66-62ac-4f85-95c3-4a54ada63941

binary-koan commented on spec/controllers/admin/refund_queues_controller_spec.rb:
```diff
@@ -30,10 +35,23 @@
+      expect(props[:pagination]).to be_present
+      expect(props[:users]).not_to match_array([hash_including(id: user_1.id)])
+      expect(props[:users]).not_to match_array([hash_including(id: user_2.id)])
+      expect(props[:users]).to match_array([hash_including(id: user_3.id)])
```
```suggestion
      expect(props[:users]).to contain_exactly(hash_including(id: user_3.id))
```

AFAIK `match_array` asserts that only the given element(s) are present, so we don't need the other checks. `contain_exactly` is essentially equivalent minus the array brackets and makes the expectation clearer to me

binary-koan commented on app/controllers/concerns/admin/list_paginated_users.rb:
```diff
@@ -0,0 +1,39 @@
+          render(
+            inertia: template,
+            props: {
+              users: InertiaRails.merge do
```
So much black magic in Inertia, I wouldn't have guessed that `InertiaRails.merge` and `WhenVisible` make a pair 😅 Concise and convenient when you get to know it I guess

EmCousin replied:
They don't. It's just InertiaRails' way to [merge props](https://inertia-rails.dev/guide/merging-props) instead of overwriting them. `WhenVisible` on the other hand is nothing more than a simple IntersectionObserver that triggers a visit upon intersecting, We could use a form or a link with a partial reload and achieve the same feature, only requiring the user to click instead of scrolling 🙂

EmCousin commented:
@binary-koan I've added a comment to explain why 5 users by default, used contain_exactly in tests (TIL!) and added the `preserveUrl` option in the paginated loader so we don't see the URL updating. Could you check again when you have the chance? :pray:
